temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+            # If the activity is done, nothing we can do but log
+            if activity.done:
+                logger.exception(
+                    f"Failed recording heartbeat (activity already done, cannot error): {err}",
+                    extra={"error": err},
+                )
+            else:
+                logger.warning(
+                    f"Cancelling activity because failed recording heartbeat: {err}",
+                    extra={"error": err},
+                    exc_info=True,
+                )
+                activity.cancel(cancelled_due_to_heartbeat_error=err)
+
+    async def _run_activity(
+        self,
+        task_token: bytes,
+        start: temporalio.bridge.proto.activity_task.Start,
+        running_activity: _RunningActivity,
+    ) -> None:
+        logger.debug("Running activity %s (token %s)", start.activity_type, task_token)
+        # We choose to surround interceptor creation and activity invocation in
+        # a try block so we can mark the workflow as failed on any error instead
+        # of having error handling in the interceptor
+        completion = temporalio.bridge.proto.ActivityTaskCompletion(
+            task_token=task_token
+        )
+        try:
+            # Find activity or fail
+            activity_def = self._activities.get(start.activity_type)
+            if not activity_def:
+                activity_names = ", ".join(sorted(self._config["activities"].keys()))
+                raise temporalio.exceptions.ApplicationError(
+                    f"Activity function {start.activity_type} is not registered on this worker, available activities: {activity_names}",
+                    type="NotFoundError",
+                    non_retryable=True,
+                )
+
+            # We must mark the running activity as sync if it's not async so
+            # that cancellation doesn't attempt to cancel the task
+            if not inspect.iscoroutinefunction(activity_def.fn):
+                running_activity.sync = True
+            else:
+                # We have to set the async form of events
+                running_activity.cancelled_event.async_event = asyncio.Event()
+                running_activity.worker_shutdown_event.async_event = asyncio.Event()
+
+            # As a special case, if we're neither a coroutine or a thread pool
+            # executor, we use a shared state manager event so it can span
+            # processes just in case they are using some kind of multiprocess
+            # executor
+            if not inspect.iscoroutinefunction(activity_def.fn) and not isinstance(
+                self._config["activity_executor"], concurrent.futures.ThreadPoolExecutor
+            ):
+                manager = self._config["shared_state_manager"]
+                # Pre-checked on worker init
+                assert manager
+                # Use cross-process events
+                running_activity.cancelled_event.thread_event = manager.new_event()
```

Comment:
Shouldn't these be set if activity is meant to be run in a threadpool executor?

Reply:
No, thread pool executor can use the existing `threading.Event` that is already there which is much cheaper. We don't even require a "shared state manager" for thread pool executor since we don't need pickle-safe eventing.

Reply:
Oh so we create those events in the constructor and replace them here?

Reply:
Yes, was just me being lazy to not ever wanting that event to be optional from a type perspective (the threading event should be cheap and is only replaced in non-async, non-thread-pool situations which shouldn't be expensive). I can change this.

Reply:
Changed.